src/backend/distributed/transaction/relation_access_tracking.c

```diff
@@ -524,10 +630,389 @@ GetRelationAccessMode(Oid relationId, ShardPlacementAccessType accessType)
+													&conflictingAccessType))
+	{
+		char *relationName = get_rel_name(relationId);
+		char *conflictingRelationName = get_rel_name(conflictingReferencingRelationId);
+
+		char *accessTypeText = PlacementAccessTypeToText(accessType);
+		char *conflictingAccessTypeText =
+			PlacementAccessTypeToText(conflictingAccessType);
+
+		if (ParallelQueryExecutedInTransaction())
+		{
+			/*
+			 * If there has already been a parallel query executed, the sequential mode
+			 * would still use the already opened parallel connections to the workers,
+			 * thus contradicting our purpose of using sequential mode.
+			 */
+			ereport(ERROR, (errmsg("cannot execute parallel %s on relation \"%s\" "
+								   "since it holds conflicting locks on reference "
+								   "relation \"%s\" due to a previous %s access "
+								   "in the same transaction", accessTypeText,
+								   relationName, conflictingRelationName,
+								   conflictingAccessTypeText),
+							errdetail("attempt to set query execution mode to sequential "
+									  "failed since a parallel query has already been "
+									  "executed in the same transaction"),
+							errhint("Try re-running the transaction with "
+									"\"SET LOCAL citus.multi_shard_modify_mode TO "
+									"\'sequential\';\"")));
+		}
+		else
+		{
+			ereport(DEBUG1, (errmsg("switching to sequential query execution mode"),
+							 errdetail(
+								 "cannot execute parallel %s on relation \"%s\" since it "
+								 "holds conflicting locks on reference relation \"%s\" due to "
+								 "a previous %s access in the same transaction ",
+								 accessTypeText, relationName, conflictingRelationName,
+								 conflictingAccessTypeText)));
+
+			SetLocalMultiShardModifyModeToSequential();
+		}
+	}
+}
+
+
+/*
+ * CheckConflictingParallelCopyAccesses is mostly a wrapper around
+ * HoldsConflictingLockWithReferencedRelations().  We're only interested in parallel
+ * accesses to distributed tables that refers reference tables via foreign constraint.
+ * Since COPY cannot be used in sequential mode, we're erroring out.
+ */
+void
+CheckConflictingParallelCopyAccesses(Oid relationId)
+{
+	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(relationId);
+	Oid conflictingReferencingRelationId = InvalidOid;
+	ShardPlacementAccessType conflictingAccessType = PLACEMENT_ACCESS_SELECT;
+
+	if (!(cacheEntry->partitionMethod == DISTRIBUTE_BY_HASH &&
+		  cacheEntry->referencedRelationsViaForeignKey != NIL))
+	{
+		return;
+	}
+
+
+	if (HoldsConflictingLockWithReferencedRelations(relationId,
+													&conflictingReferencingRelationId,
+													&conflictingAccessType))
+	{
+		char *relationName = get_rel_name(relationId);
+		char *conflictingRelationName = get_rel_name(conflictingReferencingRelationId);
+
+		char *conflictingAccessTypeText =
+			PlacementAccessTypeToText(conflictingAccessType);
+
+		ereport(ERROR, (errmsg("cannot execute parallel COPY on relation "
```

Comment:
I'd omit the conflicting locks bit, since this is a fairly deep down implementation detail. We know that this only becomes a problem in the modification case, so we can say that.

e.g. 
```
ERROR: cannot execute COPY on relation "on_update_fkey_table, since it has a foreign key to "referece_table" and "referece_table" has been modified in this transaction
```